File: sdks/android/library/src/main/java/org/xmtp/android/library/XMTPDebugInformation.kt
package org.xmtp.android.library

import uniffi.xmtpv3.FfiApiStats
import uniffi.xmtpv3.FfiIdentityStats
import uniffi.xmtpv3.FfiXmtpClient

class XMTPDebugInformation(
    private val ffiClient: FfiXmtpClient,
) {
    val apiStatistics: ApiStats
        get() = ApiStats(ffiClient.apiStatistics())
    val identityStatistics: IdentityStats
        get() = IdentityStats(ffiClient.apiIdentityStatistics())
    val aggregateStatistics: String
        get() = ffiClient.apiAggregateStatistics()

    fun clearAllStatistics() = ffiClient.clearAllStatistics()
}

class ApiStats(
    private val apiStats: FfiApiStats,
) {
    val publish: Long
        get() = apiStats.publish.toLong()
    val query: Long
        get() = apiStats.query.toLong()
    val queryNewest: Long
        get() = apiStats.queryNewest.toLong()
    val get: Long
        get() = apiStats.get.toLong()
    val subscribe: Long
        get() = apiStats.subscribe.toLong()
    val subscribeStatic: Long
        get() = apiStats.subscribeStatic.toLong()
}

class IdentityStats(
    private val identityStats: FfiIdentityStats,
) {
    val getInboxIds: Long
        get() = identityStats.getInboxIds.toLong()
    val verifySmartContractWalletSignatures: Long
        get() = identityStats.verifySmartContractWalletSignatures.toLong()
}

File: sdks/ios/Sources/XMTPiOS/Libxmtp/XMTPDebugInformation.swift
//
//  XMTPDebugInformation.swift
//  XMTPiOS
//
//  Created by Cameron Voell on 5/23/25.
//

import Foundation

public class XMTPDebugInformation {
	private let ffiClient: FfiXmtpClient

	public init(ffiClient: FfiXmtpClient) {
		self.ffiClient = ffiClient
	}

	public var apiStatistics: ApiStats {
		ApiStats(apiStats: ffiClient.apiStatistics())
	}

	public var identityStatistics: IdentityStats {
		IdentityStats(identityStats: ffiClient.apiIdentityStatistics())
	}

	public var aggregateStatistics: String {
		ffiClient.apiAggregateStatistics()
	}

	public func clearAllStatistics() {
		ffiClient.clearAllStatistics()
	}

	@available(*, deprecated, message: "uploadDebugInformation has been removed from libxmtp")
	public func uploadDebugInformation(serverUrl _: String? = nil) async throws -> String {
		// uploadDebugArchive has been removed from FFI
		throw ClientError.creationError("uploadDebugInformation is no longer available")
	}
}

public class ApiStats {
	private let apiStats: FfiApiStats

	public init(apiStats: FfiApiStats) {
		self.apiStats = apiStats
	}

	public var publish: Int64 {
		Int64(apiStats.publish)
	}

	public var query: Int64 {
		Int64(apiStats.query)
	}

	public var queryNewest: Int64 {
		Int64(apiStats.queryNewest)
	}

	public var subscribe: Int64 {
		Int64(apiStats.subscribe)
	}

	public var subscribeStatic: Int64 {
		Int64(apiStats.subscribeStatic)
	}
}

public class IdentityStats {
	private let identityStats: FfiIdentityStats

	public init(identityStats: FfiIdentityStats) {
		self.identityStats = identityStats
	}

	public var getInboxIds: Int64 {
		Int64(identityStats.getInboxIds)
	}

	public var verifySmartContractWalletSignatures: Int64 {
		Int64(identityStats.verifySmartContractWalletSignatures)
	}
}
